src/core/middlewares/vulnerable/handler.go

```diff
@@ -37,44 +41,100 @@ func New(next http.Handler) http.Handler {
+	if !ok || len(img.Digest) == 0 {
 		vh.next.ServeHTTP(rw, req)
 		return
 	}
+
+	// Is vulnerable policy set?
 	projectVulnerableEnabled, projectVulnerableSeverity, wl := util.GetPolicyChecker().VulnerablePolicy(img.ProjectName)
 	if !projectVulnerableEnabled {
 		vh.next.ServeHTTP(rw, req)
 		return
 	}
-	vl, err := scan.VulnListByDigest(img.Digest)
+
+	// Invalid project ID
+	if wl.ProjectID == 0 {
+		err := errors.Errorf("project verification error: project %s", img.ProjectName)
+		vh.sendError(err, rw)
+
+		return
+	}
+
+	// Is scanner available?
+	yes, err := scanner.DefaultController.IsScannerAvailable(wl.ProjectID)
 	if err != nil {
-		log.Errorf("Failed to get the vulnerability list, error: %v", err)
-		http.Error(rw, util.MarshalError("PROJECT_POLICY_VIOLATION", "Failed to get vulnerabilities."), http.StatusPreconditionFailed)
+		err = errors.Wrap(err, "middleware: vulnerable handler")
+		vh.sendError(err, rw)
+
 		return
 	}
-	filtered := vl.ApplyWhitelist(wl)
-	msg := vh.filterMsg(img, filtered)
-	log.Info(msg)
-	if int(vl.Severity()) >= int(projectVulnerableSeverity) {
-		log.Debugf("the image severity: %q is higher then project setting: %q, failing the response.", vl.Severity(), projectVulnerableSeverity)
-		http.Error(rw, util.MarshalError("PROJECT_POLICY_VIOLATION", fmt.Sprintf("The severity of vulnerability of the image: %q is equal or higher than the threshold in project setting: %q.", vl.Severity(), projectVulnerableSeverity)), http.StatusPreconditionFailed)
+
+	if !yes {
+		vh.next.ServeHTTP(rw, req)
 		return
 	}
-	vh.next.ServeHTTP(rw, req)
-}
 
-func (vh vulnerableHandler) filterMsg(img util.ImageInfo, filtered scan.VulnerabilityList) string {
-	filterMsg := fmt.Sprintf("Image: %s/%s:%s, digest: %s, vulnerabilities fitered by whitelist:", img.ProjectName, img.Repository, img.Reference, img.Digest)
-	if len(filtered) == 0 {
-		filterMsg = fmt.Sprintf("%s none.", filterMsg)
+	// Get the vulnerability summary
+	artifact := &v1.Artifact{
+		NamespaceID: wl.ProjectID,
+		Repository:  img.Repository,
+		Tag:         img.Reference,
+		Digest:      img.Digest,
+		MimeType:    v1.MimeTypeDockerArtifact,
 	}
-	for _, v := range filtered {
-		filterMsg = fmt.Sprintf("%s ID: %s, severity: %s;", filterMsg, v.ID, v.Severity)
+
+	cve := report.CVESet(wl.CVESet())
+	summaries, err := sc.DefaultController.GetSummary(
+		artifact,
+		[]string{v1.MimeTypeNativeReport},
+		report.WithCVEWhitelist(&cve),
+	)
+
+	if err != nil {
+		err = errors.Wrap(err, "middleware: vulnerable handler")
+		vh.sendError(err, rw)
+
+		return
 	}
-	return filterMsg
+
+	rawSummary, ok := summaries[v1.MimeTypeNativeReport]
+	// No report yet?
+	if !ok {
+		err = errors.Errorf("no scan report existing for the artifact: %s:%s@%s", img.Repository, img.Reference, img.Digest)
+		vh.sendError(err, rw)
+
+		return
+	}
+
+	summary := rawSummary.(*vuln.NativeReportSummary)
+
+	// Do judgement
+	if summary.Severity.Code() >= projectVulnerableSeverity.Code() {
```

Comment:
Here may be a problem, if the severity is `unknown` or "N/A", the `Code()` returns -1 and the image can be pulled, maybe we should see `unknown` or `N/A` a more severe vulnerability, even higher than `Critical`.